### PR TITLE
Use ruff for linting & formatting, add dev environment session

### PR DESCRIPTION
**Pull request by grahamalama**
This PR:
- replaces `flake8` with `ruff` for linting
- adds a `format` nox session to automate code formatting (and runs that formatting rule)
- adds a `dev` nox session to set up virtual environments in each microservice
  - this combined with this [Python Envy](https://marketplace.visualstudio.com/items?itemName=teticio.python-envy) VSCode extension made my development environment a bit better
  - but, perhaps this project warrants a dedicated monorepo build tool like [pants](https://www.pantsbuild.org/) 🤔

### CODE REVIEW

File: data-generator/generate_data_test.py [hunk 1]
```diff
@@ -12,14 +12,12 @@
...
-import pytest
 import datetime
-
 from urllib.request import Request
 
-from util_compare_dicts import compare_dicts
-
 import generate_data
+import pytest
+from util_compare_dicts import compare_dicts
```
Comment (hackebrot) [inline data-generator/generate_data_test.py:20]: Looks like `isort` wasn't able to categorize here.

`generate_data` and `util_compare_dicts` are local modules.

File: data-generator/generate_data_test.py [hunk 2]
```diff
@@ -205,16 +200,21 @@ def test_request(valid_change_request, make_change_request):
 def test_all_changesets_linked_with_before_attribute(generate_all_changesets):
     all_changesets = generate_all_changesets
     for i in range(1, len(all_changesets)):
-        prev_change_sha = (all_changesets[i - 1].get("checkout_sha")
-                           or all_changesets[i - 1].get("head_commit", {}).get("id"))
+        prev_change_sha = all_changesets[i - 1].get("checkout_sha") or all_changesets[
+            i - 1
+        ].get("head_commit", {}).get("id")
```
Comment (hackebrot) [inline data-generator/generate_data_test.py:205]: Not a fan of how `ruff` formats these `or` expressions, but we can fix that manually by using additional local variables.

See also [data-generator/generate_data.py](https://github.com/mozilla-services/fourkeys/pull/77/files#diff-a4086723f26c48922af02e59c87620eee44ed27a76c3dc8c775983d871e3da15)

File: noxfile.py [hunk 3]
```diff
@@ -103,25 +103,34 @@ def _determine_local_import_names(start_dir):
...
-        "--application-import-names",
-        ",".join(local_names),
-        ".",
-    ]
-    session.run("flake8", *args)
+
+@nox.session(python=False)
+@nox.parametrize("folder", FOLDERS)
+def dev(session: nox.Session, folder) -> None:
+    """Sets up a python development environment."""
```
Comment (hackebrot) [inline noxfile.py:123]:
```suggestion
    """Create Python virtual environments in subdirectories containing tests ."""
```

File: noxfile.py [hunk 4]
```diff
@@ -103,25 +103,34 @@ def _determine_local_import_names(start_dir):
...
-    args = FLAKE8_COMMON_ARGS + [
-        "--application-import-names",
-        ",".join(local_names),
-        ".",
-    ]
-    session.run("flake8", *args)
+
+@nox.session(python=False)
+@nox.parametrize("folder", FOLDERS)
+def dev(session: nox.Session, folder) -> None:
```
Comment (hackebrot) [inline noxfile.py:122]: It would be good to document that this command exists.

**Review by hackebrot — CHANGES_REQUESTED** [review]
Thanks for this pull request, @grahamalama! 🚀 

I've added my comments inline.

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+[tool.ruff.lint]
+# Default rules plus isort
```
[the remark this answers is not in the document]
Comment (grahamalama) [reply pyproject.toml]: Sure thing [19de330](https://github.com/mozilla-services/fourkeys/pull/77/commits/19de330369f857a3218b7c1037e85db248c9eb42)

File: data-generator/generate_data_test.py (hunk 1 above)
Comment (grahamalama) [reply data-generator/generate_data_test.py:21]: Ahh yeah, had to specify a [`src` option](https://docs.astral.sh/ruff/settings/#src):

[14ecb94](https://github.com/mozilla-services/fourkeys/pull/77/commits/14ecb94a68cce1513d58fc05f869928a813b3826), [e09ddf3](https://github.com/mozilla-services/fourkeys/pull/77/commits/e09ddf3016fb87ecb9edbd3ec7280f956f0a1591)

File: data-generator/generate_data_test.py (hunk 2 above)
Comment (grahamalama) [reply data-generator/generate_data_test.py]: [d536b05](https://github.com/mozilla-services/fourkeys/pull/77/commits/d536b0558c27aa6e26ebb614bbe7aebf0a523c13)

File: noxfile.py (hunk 4 above)
Comment (grahamalama) [reply noxfile.py:103]: [daa6320](https://github.com/mozilla-services/fourkeys/pull/77/commits/daa63204bc59cef7949170b362d04a2d056e42bf)

File: .git-blame-ignore-revs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Run: git config blame.ignoreRevsFile .git-blame-ignore-revs
+# to remove the following commits from git-blame
+
+# Run `ruff` formatting after adding to noxfile
+7e782b33d262476b2c33e9b9843438270455a0fe
```
Comment (grahamalama) [inline .git-blame-ignore-revs:5]: TODO: update this SHA after receiving and approval and rebasing

**Comment by grahamalama** [comment]
@hackebrot pretty sure I addressed all your comments. Let me know what else you have.

**Review by hackebrot — APPROVED** [review]
LGTM! 👍🏻

**Comment by grahamalama** [comment]
Thanks for the approval @hackebrot. Barring any objections, going to force push my [rebased branch](https://github.com/mozilla-services/fourkeys/compare/main...format-and-dev-rebase) here then merge with a merge commit

**Pull request merged**